### PR TITLE
select ResidentCompiler during FlutterDevice initialization

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Rather than have each command be responsible for creating the correct ResidentCompiler, use a static method to create the FlutterDevice which is responsible for creating it. This simplifies the commands responsible for creating the flutter device, and frees us up to use different generators/build scripts for flutter web or other targets.

This also moves the generation of the .packages file outside of the builders. In cases where the build fails we still to produce a valid .packages file. This also makes it easier for flutter tester

## Related Issues

https://github.com/flutter/flutter/issues/27141

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_runner/build_runner.dart [hunk 1]
```diff
@@ -265,19 +270,28 @@ class BuildRunner extends CodeGenerator {
...
     }));
     final String relativeMain = fs.path.relative(mainPath, from: flutterProject.directory.path);
-    final File generatedPackagesFile = fs.file(fs.path.join(flutterProject.generated.path, fs.path.setExtension(relativeMain, '.packages')));
     final File generatedDillFile = fs.file(fs.path.join(flutterProject.generated.path, fs.path.setExtension(relativeMain, '.app.dill')));
-    return _BuildRunnerCodegenDaemon(buildDaemonClient, generatedPackagesFile, generatedDillFile);
+    return _BuildRunnerCodegenDaemon(buildDaemonClient, generatedDillFile);
+  }
+
+  // Create generated packages file which adds a multi-root scheme to the user's
+  // project directory. currently we only replace the root package with a multiroot
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/build_runner.dart]: nit: "Currently"

**Comment by jonahwilliams** [comment]
PTAL @yjbanov

File: packages/flutter_tools/lib/src/compile.dart [hunk 2]
```diff
@@ -662,7 +662,7 @@ class ResidentCompiler {
   }
 
   Future<dynamic> shutdown() {
-    _server.kill();
-    return _server.exitCode;
+    _server?.kill();
+    return _server?.exitCode;
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/compile.dart]: Is it meaningful to return `null` if `_server` is `null`? Maybe return `0`? If it an unexpected situation, perhaps throw or print a warning?

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/compile.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart]: Done

File: packages/flutter_tools/lib/src/build_runner/build_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/build_runner.dart]: Done

File: packages/flutter_tools/lib/src/build_runner/build_runner.dart [hunk 3]
```diff
@@ -21,12 +21,15 @@ import '../base/process_manager.dart';
...
 import '../codegen.dart';
 import '../convert.dart';
+import '../dart/package_map.dart';
 import '../dart/pub.dart';
 import '../globals.dart';
 import '../project.dart';
 import '../resident_runner.dart';
 import 'build_script_generator.dart';
 
+const String _kMultirootScheme = 'org-dartlang-app';
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/build_runner.dart]: Done

File: packages/flutter_build/lib/src/kernel_builder.dart [hunk 4]
```diff
@@ -89,25 +89,18 @@ class FlutterKernelBuilder implements Builder {
...
     final AssetId outputId = buildStep.inputId.changeExtension(_kFlutterDillOutputExtension);
-    final AssetId packagesOutputId = buildStep.inputId.changeExtension(_kPackagesExtension);
 
     // Create a scratch space file that can be read/written by the frontend server.
     // It is okay to hard-code these file names because we will copy them back
     // from the temp directory at the end of the build step.
     final Directory tempDirecory = await Directory.systemTemp.createTemp('_flutter_build');
-    final File packagesFile = File(path.join(tempDirecory.path, _kPackagesExtension));
+    final Directory projectDir = File(packagesPath).parent;
+    final String packagesFilePath = path.join(projectDir.path, '.packages.generated');
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_build/lib/src/kernel_builder.dart]: I'm going to delete this file in the next PR, so for the sake of brevity I'll leave it as is for now.

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**